File: fontes/br.com.wdc.kt.shopping/br.com.wdc.kt.shopping.view.react.client/src/scripts/utils/RSA.ts
import BigIntUtils, { type BigIntValue } from './BigIntUtils'

export default class RSA {
  static readonly N65537 = 65537n

  #publicExponent: bigint
  #privateKey: bigint
  #publicKey: bigint

  constructor(publicExponent: BigIntValue, privateKey: BigIntValue, publicKey: BigIntValue) {
    if (typeof publicExponent !== 'bigint') publicExponent = BigInt(publicExponent)
    if (typeof privateKey !== 'bigint') privateKey = BigInt(privateKey)
    if (typeof publicKey !== 'bigint') publicKey = BigInt(publicKey)

    this.#publicExponent = publicExponent
    this.#privateKey = privateKey
    this.#publicKey = publicKey
  }

  getPublicExponent() {
    return this.#publicExponent
  }

  getPublicKey() {
    return this.#publicKey
  }

  getPrivateKey() {
    return this.#privateKey
  }

  encrypt(message: BigIntValue) {
    return BigIntUtils.modPow(message, this.#publicExponent, this.#publicKey)
  }

  decrypt(encrypted: BigIntValue) {
    return BigIntUtils.modPow(encrypted, this.#privateKey, this.#publicKey)
  }

  toString() {
    return [
      '{',
      'publicExponent: ',
      this.#publicExponent.toString(16),
      ', publicKey: ',
      this.#publicKey.toString(16),
      ', private: ',
      this.#privateKey.toString(16),
      '}',
    ].join()
  }
}

File: fontes/br.com.wdc.kt.framework/br.com.wdc.kt.framework.commons/src/main/kotlin/br/com/wdc/framework/commons/security/RSA.kt
package br.com.wdc.framework.commons.security

import java.math.BigInteger
import java.util.Random

class RSA {

    var privateKey: BigInteger
        private set
    var publicExponent: BigInteger
        private set
    var publicKey: BigInteger
        private set

    constructor(n: Int, random: Random) {
        val p = BigInteger.probablePrime(n / 2, random)
        val q = BigInteger.probablePrime(n / 2, random)
        val phi = (p - BigInteger.ONE) * (q - BigInteger.ONE)

        publicKey = p * q
        publicExponent = N65537
        privateKey = publicExponent.modInverse(phi)
    }

    constructor(publicExponent: BigInteger, privateKey: BigInteger, publicKey: BigInteger) {
        this.publicExponent = publicExponent
        this.privateKey = privateKey
        this.publicKey = publicKey
    }

    fun encrypt(message: BigInteger): BigInteger = message.modPow(publicExponent, publicKey)

    fun decrypt(encrypted: BigInteger): BigInteger = encrypted.modPow(privateKey, publicKey)

    override fun toString(): String = buildString(256) {
        append("publicExponent  = ").append(publicExponent.toString(16)).append('\n')
        append("publicKey  = ").append(publicKey.toString(16)).append('\n')
        append("privateKey  = ").append(privateKey.toString(16)).append('\n')
    }

    companion object {
        @JvmField
        val N65537: BigInteger = BigInteger.valueOf(65537)
    }
}
